util/topsql/reporter/reporter.go

```diff
@@ -84,28 +85,42 @@ type RemoteTopSQLReporter struct {
 func NewRemoteTopSQLReporter(decodePlan planBinaryDecodeFunc) *RemoteTopSQLReporter {
 	ctx, cancel := context.WithCancel(context.Background())
 	tsr := &RemoteTopSQLReporter{
-		DefaultDataSinkRegisterer: NewDefaultDataSinkRegisterer(ctx),
-		ctx:                       ctx,
-		cancel:                    cancel,
-		collectCPUTimeChan:        make(chan []tracecpu.SQLCPUTimeRecord, collectChanBufferSize),
-		collectStmtStatsChan:      make(chan stmtstats.StatementStatsMap, collectChanBufferSize),
-		reportCollectedDataChan:   make(chan collectedData, 1),
-		collecting:                newCollecting(),
-		normalizedSQLMap:          newNormalizedSQLMap(),
-		normalizedPlanMap:         newNormalizedPlanMap(),
-		stmtStatsBuffer:           map[uint64]stmtstats.StatementStatsMap{},
-		decodePlan:                decodePlan,
+		ctx:                     ctx,
+		cancel:                  cancel,
+		collectCPUTimeChan:      make(chan []collector.SQLCPUTimeRecord, collectChanBufferSize),
+		collectStmtStatsChan:    make(chan stmtstats.StatementStatsMap, collectChanBufferSize),
+		reportCollectedDataChan: make(chan collectedData, 1),
+		collecting:              newCollecting(),
+		normalizedSQLMap:        newNormalizedSQLMap(),
+		normalizedPlanMap:       newNormalizedPlanMap(),
+		stmtStatsBuffer:         map[uint64]stmtstats.StatementStatsMap{},
+		decodePlan:              decodePlan,
 	}
+	tsr.sqlCPUCollector = collector.NewSQLCPUCollector(tsr)
+	tsr.DefaultDataSinkRegisterer = NewDefaultDataSinkRegisterer(ctx, tsr)
+
 	go tsr.collectWorker()
 	go tsr.reportWorker()
 	return tsr
 }
 
+// Enable implemented StateController interface.
+func (tsr *RemoteTopSQLReporter) Enable() {
+	tsr.sqlCPUCollector.Start()
+	topsqlstate.GlobalState.Enable.Store(true)
+}
+
+// Disable implemented StateController interface.
+func (tsr *RemoteTopSQLReporter) Disable() {
+	tsr.sqlCPUCollector.Stop()
+	topsqlstate.GlobalState.Enable.Store(false)
+}
```

Comment:
It seems incorrect that the external can control the top sql feature via the `tsr`. By design, the only way to control the top sql state is the reporter's internal logic. No other 3rd-parties are allowed to control.